### PR TITLE
13566 get all ei

### PR DESCRIPTION
**Pull request by abuts**
This solves #13566 

Despite there are a lot of commits here, the changes themselves are not that big. It was unclear from the beginning what should be done so more then three iterations have been performed to achieve result I consider satisfactory. Most changes were done more then once. 

Here we have an algorithm, which analyses chopper logs and signal on monitors to return list of incident energies used in direct inelastic experiment. To achieve this, 4 things have been added to Mantid:

1) An auxiliary method in kernel which performs the running averages in an interval over vector with heterogeneous binning and unit test for it.
2) A auxiliary method in time series property, which returns derivative of this property with unit test for the method. 
3) chopper-position component added to ISIS direct instruments to provide information about chopper logs and the way to analyse them. 
4) the GetAllEi algorithm itself with docs, doctest, unit and system tests for it and its components and modifications to CreateSampleWorkspace and WorkspaceCreationHelpers to allow testing GetAllEi. I hope that documentation sufficiently describes what and how the algorithm does.

I run the algorithm on wide range of recent ISIS runs and it does what it should do so tester may be more or less confident that it works, though now I see how it can be written better. 
Code review and general look through is suggested in the areas 1,2 and 4. Look how clear the algorithm description and general design is. I have downgraded from C++11 to C++(std) to support RHEL6, so there are a lot of former lambda functions in anonymous namespace. Careful tester may try algorithm on some ISIS direct runs though this is very tedious procedure I've already performed. If one going to try this, I would suggest asking me for the python code to perform these tests. It sits in ISIS local network (ndw932\Data\Mantid_Testing\15_09_03 (current IP 130.246.50.98). I have not used GUI very much so this is what may be worth trying extensively. 

Look at auxiliary methods 1 and 2. Could they be done better? My primary problem in 1) was shift in peak location due to subsequent averaging, and though this can not be generally achieved for arbitrary peaks, I believe I've achieved the decent result. Still not very satisfactory in general sense, but I believe sufficient for the algorithm purpose.

Item 3) can be done better, but error in current instrument implementation (access to chopper through Instrument interface does not return parameter map) defines current design. This may be subject to further improvements (needs discussion with Nick)

### CODE REVIEW

File: .githooks/pre-commit [hunk 1]
```diff
@@ -1,334 +0,0 @@
-#!/usr/bin/env bash
```
Comment (mantid-roman) [inline .githooks/pre-commit]: Why was this file deleted?

**Comment by abuts** [comment]
It should not.

File: Framework/Kernel/src/VectorHelper.cpp [hunk 2]
```diff
@@ -533,6 +533,166 @@ void linearlyInterpolateY(const std::vector<double> &x, std::vector<double> &y,
...
+*                     from) should be: index>=startIndex>=0
+*@param endIndex   -- index in the array of data (input to end average at)
+*                     should be: index<=endIndex<=input.size()
+*@param halfWidth  -- half width of the interval to integrate.
+*@param input      -- vector of input signal
+*@param binBndrs   -- pointer to vector of bin boundaries or NULL pointer.
+ */
+double runAverage(size_t index, size_t startIndex, size_t endIndex,
+                  const double halfWidth, const std::vector<double> &input,
+                  std::vector<double> const *const binBndrs) {
```
Comment (mantid-roman) [inline Framework/Kernel/src/VectorHelper.cpp]: Does binDirs have to be a pointer? Can it be a const ref? The function's code would look clearer without ->operator[](...)

File: Framework/Kernel/src/VectorHelper.cpp [hunk 3]
```diff
@@ -533,6 +533,166 @@ void linearlyInterpolateY(const std::vector<double> &x, std::vector<double> &y,
...
+
+  size_t iStart, iEnd;
+  double weight0(0), weight1(0), start, end;
+  //
+  if (binBndrs) {
+    // identify initial and final bins to
```
Comment (mantid-roman) [inline Framework/Kernel/src/VectorHelper.cpp]: I see

Comment (mantid-roman) [inline Framework/Kernel/src/VectorHelper.cpp]: Still

const vector<...>& bndrs = *binBndrs;

could be better

File: Framework/Kernel/src/VectorHelper.cpp [hunk 4]
```diff
@@ -533,6 +533,166 @@ void linearlyInterpolateY(const std::vector<double> &x, std::vector<double> &y,
...
+  }
+
+  size_t length = endIndex - startIndex;
+  output.resize(length);
+
+  double halfWidth = avrgInterval / 2;
+  if (!binBndrs) {
+    if (std::fabs(double(static_cast<size_t>(halfWidth)) * 2 - avrgInterval) >
+        1.e-6) {
+      halfWidth = static_cast<double>(static_cast<size_t>(halfWidth) + 1);
```
Comment (mantid-roman) [inline Framework/Kernel/src/VectorHelper.cpp]: Was your intention to do this?

halfWidth = std::floor(halfWidth) + 1.0;
